src/api/helpers.py

```diff
@@ -0,0 +1,329 @@
+    """
+    config = TABLE_CONFIG[table_name]
+    record_data = record_field.get("omop_record", {})
+
+    # 1. Extract the temporary ID object for the primary key, if it exists
+    pk_field = config.get("pk")
+    id_obj = None
+    if pk_field:
+        id_obj = record_data.pop(pk_field, None)
+
+    # 2. Resolve all foreign keys using the id_mapper
+    for fk_field_name in config["fk_map"]:
+        if fk_field_name in record_data:
+            fk_id_obj = record_data[fk_field_name]
+            actual_fk_id = id_mapper.get_id(fk_id_obj)
+            if actual_fk_id is None:
+                raise ProblemException(
+                    status=400,
+                    title="Bad Request",
+                    detail=(
+                        f"Could not resolve foreign key for '{fk_field_name}' in table '{table_name}'. "
+                        f"Ensure the referenced entity is created first. "
+                        f"Missing reference: {fk_id_obj}"
+                    ),
+                )
+            record_data[fk_field_name] = actual_fk_id
+
+    # 3. Call the specific create function for this table
+    create_function = config["create_func"]
+    new_record = await create_function(session, record_data)
+
+    # 4. Store the mapping for the new primary key
+    if pk_field and id_obj:
+        new_pk_value = new_record[pk_field]
+        id_mapper.store_id(id_obj, new_pk_value)
+
+    # 5. Prepare the record for the final return object
+    new_record["omop_table"] = table_name
+    return new_record
+
+
+# ==============================================================================
+# Insert donor with clinical data
+# ==============================================================================
+async def ingest_donor_with_clinical_data(
+    session: AsyncSession, donor_data: dict, return_obj: list
+):
+    """
+    Insert a single donor with related clinical data
+    """
+    id_mapper = IdMapper()
+    PATTERNS = {
+        "donor": r"^\$\.donors\[\d+\]$",
+        "primary_diagnosis": r"^\$\.donors\[\d+\].primary_diagnoses\[\d+\]$",
+        "specimen": r"^\$\.donors\[\d+\].primary_diagnoses\[\d+\].specimens\[\d+\]$",
+        "treatment": r"^\$\.donors\[\d+\].primary_diagnoses\[\d+\].treatments\[\d+\]$",
+        "systemic_therapy": r"^\$\.donors\[\d+\].primary_diagnoses\[\d+\].treatments\[\d+\].systemic_therapies\[\d+\]$",
+        "surgery": r"^\$\.donors\[\d+\].primary_diagnoses\[\d+\].treatments\[\d+\].surgeries\[\d+\]$",
+        "radiation": r"^\$\.donors\[\d+\].primary_diagnoses\[\d+\].treatments\[\d+\].radiations\[\d+\]$",
+        "followup": r"^\$\.donors\[\d+\].primary_diagnoses\[\d+\].treatments\[\d+\].followups\[\d+\]$",
+        "biomarker": r"^\$\.donors\[\d+\].biomarkers\[\d+\]$",
+    }
+
+    new_dataset_id = None
+    for key in donor_data.keys():
+        items = donor_data[key]
+        if re.match(PATTERNS["donor"], key):
+            # First pass: Find and create the dataset
```

Comment:
I'll note that this doesn't necessarily pass over the donors keys first; I'm unsure after reading it if it's necessary to create all donors first but if so you'll want to make sure that there's actually two passes.

Reply:
Fixed in [c6be9b9](https://github.com/CanDIG/candig-api/pull/9/commits/c6be9b92f8906dfa871a071b019678b9ff326011)